### PR TITLE
Detail attachement bug to sentry

### PR DESCRIPTION
**Pull request by NicolasBourdin88**
- Create new exception to detail bug 
- Hide file name in exception message

### CODE REVIEW

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt [hunk 1]
```diff
@@ -122,7 +163,7 @@ object LocalStorageUtils {
         if (userDir.hasNoChildren) userDir.delete()
         if (attachmentsRootDir.hasNoChildren) attachmentsRootDir.delete()
     }
-    //endregion
+//endregion
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt:166]: Revert this diff.

File: app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt [hunk 2]
```diff
@@ -136,5 +177,7 @@ object LocalStorageUtils {
     private fun generateRootDir(directory: File, userId: Int, mailboxId: Int): File {
         return File(directory, "$userId/$mailboxId")
     }
-    //endregion
+//endregion
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt:180]: Revert this diff.

File: app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt [hunk 3]
```diff
@@ -87,21 +90,59 @@ object LocalStorageUtils {
...
+            FileOutputStream(file).use(inputStream::copyTo)
+            true
+        }.getOrElse {
             Sentry.withScope { scope ->
-                scope.setExtra("uri", uri.toString())
-                Sentry.captureMessage("failed to access uri")
+                scope.setExtra("uri", uri.toString().replace(file.path, "HIDDEN FILE NAME"))
+                val exception =
+                    it.message?.let { AttachmentMissingFileException(it.replace(file.path, "HIDDEN FILE NAME")) }
+                        ?: it
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt:131]: There's 2 different `it` here, rename at least one of them _(or maybe all of them)_.

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt [hunk 4]
```diff
@@ -122,6 +124,40 @@ object LocalStorageUtils {
...
+                    ?: it
+
+                exception.stackTrace = it.stackTrace
+                Sentry.captureException(exception)
+
+                if (exception.message?.contains("ENAMETOOLONG") == true) {
+                    snackbarManager.postValue(context.getString(R.string.errorFileNameTooLong))
+                } else {
+                    snackbarManager.postValue(context.getString(RCore.string.errorFileNotFound))
+                }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt:153]: Use the IF / ELSE to get the `@IdRes`, then `postValue()` only once.

File: app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt [hunk 5]
```diff
@@ -122,6 +124,40 @@ object LocalStorageUtils {
...
+    ): File? {
+        val file = File(attachmentsUploadDir, hashedFileName)
+        val isSuccess = runCatching {
+            FileOutputStream(file).use(inputStream::copyTo)
+            true
+        }.getOrElse {
+            Sentry.withScope { scope ->
+                scope.setExtra("uri", uri.toString().replace(file.path, "HIDDEN FILE NAME"))
+                val exception = it.message
+                    ?.let { message -> AttachmentMissingFileException(message.replace(file.path, "HIDDEN FILE NAME")) }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt:143]: Duplicated & hardcoded string.
Move it to const.

File: app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt [hunk 6]
```diff
@@ -122,6 +124,40 @@ object LocalStorageUtils {
...
+                    ?: it
+
+                exception.stackTrace = it.stackTrace
+                Sentry.captureException(exception)
+
+                if (exception.message?.contains("ENAMETOOLONG") == true) {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt:149]: Hardcoded string, move it to const.

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt [hunk 7]
```diff
@@ -122,6 +126,39 @@ object LocalStorageUtils {
...
+                    ?.let { message -> AttachmentMissingFileException(message.replace(file.path, HIDDEN_FILE_NAME)) }
+                    ?: it
+
+                exception.stackTrace = it.stackTrace
+                Sentry.captureException(exception)
+
+                val isNameLong = exception.message?.contains(NAME_TOO_LONG_EXCEPTION) == true
+                val snackbarMessageId = if (isNameLong) R.string.errorFileNameTooLong else RCore.string.errorFileNotFound
+
+                snackbarManager.postValue(context.getString(snackbarMessageId))
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt:154]:
```suggestion
                val errorId = if (exception.message?.contains(NAME_TOO_LONG_EXCEPTION) == true) {
                    R.string.errorFileNameTooLong
                } else {
                    RCore.string.errorFileNotFound
                }
                snackbarManager.postValue(context.getString(errorId))
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt [hunk 8]
```diff
@@ -103,14 +108,13 @@ object LocalStorageUtils {
...
+        snackbarManager: SnackbarManager,
     ): File? {
         return context.contentResolver.openInputStream(uri)?.use { inputStream ->
             val attachmentsUploadDir = getAttachmentUploadDir(context, draftLocalUuid, attachmentLocalUuid)
             attachmentsUploadDir.mkdirs()
             val hashedFileName = "${uri.toString().substringAfter("document/").hashCode()}_$fileName"
-            File(attachmentsUploadDir, hashedFileName).also { file ->
-                FileOutputStream(file).use(inputStream::copyTo)
-            }
+            return@use getFileToUpload(context, uri, snackbarManager, attachmentsUploadDir, hashedFileName, inputStream)
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/utils/LocalStorageUtils.kt:117]:
```suggestion

            return@use getFileToUpload(context, uri, snackbarManager, attachmentsUploadDir, hashedFileName, inputStream)
```

**Review by FabianDevel — CHANGES_REQUESTED** [review]

**Review by FabianDevel — APPROVED** [review]

**Pull request merged**